python/tchannel/tornado/connection.py

```diff
@@ -618,21 +611,19 @@ def send_request(self, request):
         return response_future
 
     def adapt_result(self, f, request, response_future):
+        if not response_future.running():
+            return
+
         if f.exception():
             protocol_exception = f.exception()
             protocol_exception.tracing = request.tracing
             response_future.set_exception(protocol_exception)
-            # event: after_receive_protocol_error
-            self.tchannel.event_emitter.fire(
-                EventType.after_receive_error,
-                protocol_exception,
-            )
+
         else:
             response = f.result()
             response.tracing = request.tracing
             response_future.set_result(response)
-            # event: after_receive_response
-            self.tchannel.event_emitter.fire(
-                EventType.after_receive_response,
-                response,
-            )
+
+    def remove_outstanding_request(self, request):
+        """Remove request from pending request list"""
+        self._outstanding.pop(request.id, None)
```

Comment:
A precondition to doing this is that the request has finished, right? That is, the future has stopped running. If that turns out to not be the case, is that a programmer error? Maybe you want an assert statement here before doing the pop? Make sure that `future.running()` is False?